### PR TITLE
Class view should show nested private types if source code is available.

### LINKED ISSUE #21329: Class View should show non-public nested types.
**Version Used**: 

Visual Studio 2017 Community (v15.2)

**Steps to Reproduce**:

Type the following VB code into any kind of project (e.g. a console project)

```vb
Public Class OuterClass
  Public Class PublicInnerClass
  End Class

  Private Class PrivateInnerClass
  End Class

  Protected Class ProtectedInnerClass
  End Class
End Class
```

...or the following C# code:

```csharp
  class OuterClass
  {
    public class PublicInnerClass { }
    private class PrivateInnerClass { }
    protected class ProtectedInnerClass { }

  }
```

Now open Class View directly (or put the caret on `PrivateInnerClass` and execute the `View.SynchronizeClassView` command, e.g. from the Immediate window).

**Expected Behavior**:

The Class View should show all classes (the outer class and all the nested inner classes with different visibility modifiers). If you use the `View.SynchronizeClassView` command on a non-public nested class, that should also jump to said class.

**Actual Behavior**:

Class View only shows `OuterClass` and `OuterClass.PublicInnerClass`. It does not show the `PrivateInnerClass` and `ProtectedInnerClass` and as a result, the `View.SynchronizeClassView` command also fails to navigate to those non-public nested classes. Ticking all the various "Show..." options in that Class View gear icon doesn't help.

I think the problem is that the nested types are non-public, because the same problem happens if you try this with non-public enum/struct/interface types.

I say this is a bug because if Class View can show me `OuterClass` for which I have the source code, it makes no sense to not show me `PrivateInnerClass` and `ProtectedInnerClass` as they are (by definition) also available in the same source code.

### PR DESCRIPTION
**Pull request by MaStr11**
**Customer scenario**

Class view should show nested private types if source code is available. See #21329 for the motivation of the request. 

Implementation remarks:
The new method `IncludeTypeMember` has the about the same structure as the existing method `IncludeMemberSymbol`. Most changes are related to the tests.

**Bugs this fixes:**

Fixes #21329

**Workarounds, if any**

None.

**Risk**

Might break class view or shows types in class view that are not relevant to the user.

**Performance impact**

Low.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Private type members were considered but private types were not.

**How was the bug found?**

Customer reported.

**Test documentation updated?**

Maybe https://msdn.microsoft.com/en-us/library/whks02wf.aspx needs to be adopted.

### CODE REVIEW

**Comment by Pilchie** [comment]
Thanks for the PR @MaStr11!

Tagging @dotnet/roslyn-ide and @DustinCampbell for review here.

File: src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractListItemFactory.cs [hunk 1]
```diff
@@ -589,6 +589,26 @@ private ImmutableArray<INamedTypeSymbol> GetAccessibleTypeMembers(INamespaceOrTy
             return builder.ToImmutable();
         }
 
+        private bool IncludeTypeMember(INamedTypeSymbol typeMember, IAssemblySymbol assemblySymbol)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractListItemFactory.cs]: 💭 There are two ways to approach this:

* If `typeMember.Locations.Any(l => l.IsInSource)` is "faster", then the current approach is good
* If `typeMember.IsAccessibleWithin(assemblySymbol)` is "faster", then the following approach might be better:

```csharp
if (!IncludeSymbol(typeMember))
{
  return false;
}

if (typeMember.IsAccessibleWithin(assemblySymbol))
{
  return true;
}

return typeMember.Locations.Any(l => l.IsInSource);
```

:memo: Based on the form of the original code, I lean towards the second approach as the safer option.

Comment (MaStr11) [reply src/VisualStudio/Core/Def/Implementation/Library/ObjectBrowser/AbstractListItemFactory.cs]: `IsAccessibleWithin` ends up calling [IsSymbolAccessibleCore](https://github.com/MaStr11/roslyn/blob/ClassViewShouldShowNonPublicNestedTypes/src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions_Accessibility.cs#L79). `IsSymbolAccessibleCore` is checking also cross assembly visibility. So this is heavy. `typeMember.Locations.Any(l => l.IsInSource)` is an lookup in an ImmutableArray and IsInSource is just `bool IsInSource { get { return SourceTree != null; } }`. So performance-wise `Locations.Any(l => l.IsInSource);` should be faster than `IsAccessibleWithin`.

**Comment by sharwell** [comment]
@MaStr11 Based on the result in #22580, I'm marking this as **blocked** until we can find a more efficient way to determine if a symbol is located in source.

**Comment by MaStr11** [comment]
@sharwell Given that #23491 is merged I assumed that this PR can be unblocked.

**Comment by sharwell** [comment]
@dotnet-bot retest this please

**Comment by sharwell** [comment]
@Pilchie What milestone should this target?

**Comment by Pilchie** [comment]
I would be fine with either 15.7 or 15.8.

**Comment by sharwell** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please

**Comment by ericmutta** [comment]
@MaStr11 many thanks for this fix, you are a star! 😄 

And equally many thanks to @sharwell @Pilchie @DustinCampbell for accepting it! 👍

**Pull request merged**